### PR TITLE
Prompt before reducing number of rows

### LINKED ISSUE #13952: Insert page should prompt before reducing number of rows
On the Insert page, when we add rows to the page with the "Continue insertion with __ rows" function, there is a risk of data loss if the user adds data to these rows then accidentally reduces the number of rows. When changing the number of rows, we should check if we're reducing the number, and if we are then we should check if there's been data input and prompt the users.

### PR DESCRIPTION
**Pull request by shucon**
Fixes: https://github.com/phpmyadmin/phpmyadmin/issues/13952

Signed-off-by: Saksham Gupta <shucon01@gmail.com>

Before submitting pull request, please check that every commit:

- [x] Has proper Signed-Off-By
- [x] Has commit message which describes it
- [x] Is needed on it's own, if you have just minor fixes to previous commits, you can squash them
- [ ] Any new functionality is covered by tests

### CODE REVIEW

File: js/messages.php [hunk 1]
```diff
@@ -69,6 +69,8 @@ function () {
     = __('Do you really want to delete the search "%s"?');
 $js_messages['strConfirmNavigation']
     = __('You have unsaved changes; are you sure you want to leave this page?');
+$js_messages['strConfirmRowChange']
+    = __('You have unsaved changes; are you sure you want to reduce number of rows?');
```
Comment (ibennetch) [inline js/messages.php]: Your message is okay, but I've been trying to make error messages and warnings more user-friendly. How does "You are trying to reduce the number of rows, but have already entered data in those rows which will be lost. Do you wish to continue?" sound to everyone?

**Comment by ibennetch** [comment]
I haven't tested it but this looks quite good.

File: js/ajax.js [hunk 2]
```diff
@@ -727,6 +750,17 @@ AJAX.registerOnload('functions.js', function () {
...
         AJAX.resetLock();
     });
+    /**
+     * Displays warning if number of rows are reduced
+     * and chances of data loss.
+     */
+    var row = document.getElementById('insert_rows');
+    if (row !== null) {
+        var row_count = Number(row.value);
+        row.addEventListener("change", function(){
```
Comment (mauriciofauth) [inline js/ajax.js]: An event listener already exists for this selector. Why not use it?

https://github.com/phpmyadmin/phpmyadmin/blob/ef99fef8d6b760a2f3552c89afe50ba6d5562add/js/tbl_change.js#L463

File: js/ajax.js [hunk 3]
```diff
@@ -508,6 +508,27 @@ var AJAX = {
...
+
+        if (newCount < rowCount && !jQuery.isEmptyObject(AJAX.lockedTargets)
+            && confirm(PMA_messages.strConfirmRowChange) === true
+        ) {
+            return newCount;
+        } else {
+            return newCount;
+        }
+    },
     /**
```
Comment (devenbansod) [inline js/ajax.js]: @shucon I'm not sure if this file would be an ideal place to add this function. It's not related to AJAX as such, more related to a specific page.

Comment (shucon) [reply js/ajax.js]: @devenbansod I did this for two reasons.
1. A similar function for the same page is written here (line 249).
2. I had to use AJAX.lockedTargets . Although I could use it anywhere but writing it here seemed right because of the above reason.

File: js/ajax.js [hunk 4]
```diff
@@ -508,6 +508,21 @@ var AJAX = {
             }
         }
     },
+    /**
+     * Displays alert if data loss possible on decrease
+     * of rows.
+     */
+    rowCheck: function (row, rowCount) {
+        var newCount = Number(row.value);
+        if (newCount < rowCount && !jQuery.isEmptyObject(AJAX.lockedTargets)
```
Comment (mauriciofauth) [inline js/ajax.js]: You don't need to check if the number of rows the user entered is less than or greater than the number of rows on the page. This check already exists, you only need to add the confirmation message.

File: js/tbl_change.js [hunk 5]
```diff
@@ -460,7 +460,17 @@ AJAX.registerOnload('tbl_change.js', function () {
...
+    if (row !== null) {
+        var rowCount = Number(row.value);
+    }
     $(document).on('change', '#insert_rows', function (event) {
+
+        /**
+         * Displays warning if number of rows are reduced
+         * and chances of data loss.
+         */
+        rowCount = AJAX.rowCheck(row, rowCount);
```
Comment (mauriciofauth) [inline js/tbl_change.js]: You are returning a value, but you are not using it for anything.

**Review by MauricioFauth — CHANGES_REQUESTED** [review]

File: js/ajax.js (hunk 3 above)
Comment (mauriciofauth) [reply js/ajax.js]: I have to agree with Deven.

File: js/ajax.js (hunk 4 above)
Comment (shucon) [reply js/ajax.js]: Where does this check exists? Also , I would have to provide some condition to show the confirmation message.

File: js/tbl_change.js (hunk 5 above)
Comment (shucon) [reply js/tbl_change.js]: I'm using the returned value to update the old row count and compare it with the new count that the user wants to check the rowCheck condition.

File: js/ajax.js (hunk 3 above)
Comment (shucon) [reply js/ajax.js]: Should I transfer the code to tbl_change.js ?

File: js/ajax.js (hunk 4 above)
Comment (mauriciofauth) [reply js/ajax.js]: This check is already done when adding or removing lines from the page.

https://github.com/phpmyadmin/phpmyadmin/blob/ef99fef8d6b760a2f3552c89afe50ba6d5562add/js/tbl_change.js#L483

https://github.com/phpmyadmin/phpmyadmin/blob/ef99fef8d6b760a2f3552c89afe50ba6d5562add/js/tbl_change.js#L669

Comment (shucon) [reply js/ajax.js]: Thanks @mauriciofauth I've done the required changes.

File: js/tbl_change.js [hunk 6]
```diff
@@ -667,12 +672,22 @@ AJAX.registerOnload('tbl_change.js', function () {
...
-                    .addBack()
-                    .remove();
-                curr_rows--;
+            /**
+             * Displays alert if data loss possible on decrease
+             * of rows.
+             */
+            if (jQuery.isEmptyObject(AJAX.lockedTargets) || (!jQuery.isEmptyObject(AJAX.lockedTargets)
+                && confirm(PMA_messages.strConfirmRowChange) === true)
+            ) {
```
Comment (mauriciofauth) [inline js/tbl_change.js]: Don't perform complex actions in if statements. Put this jQuery call into a variable.

This second jQuery call is redundant and can be removed. The right side of the `||` will only run if the left side is false.

File: js/tbl_change.js [hunk 7]
```diff
@@ -667,12 +672,22 @@ AJAX.registerOnload('tbl_change.js', function () {
...
+                while (curr_rows > target_rows) {
+                    $('input[id^=insert_ignore]:last')
+                        .nextUntil('fieldset')
+                        .addBack()
+                        .remove();
+                    curr_rows--;
+                }
+            } else {
+                    document.getElementById('insert_rows').value = curr_rows;
```
Comment (mauriciofauth) [inline js/tbl_change.js]: Fix the indentation of this block.

**Review by MauricioFauth — CHANGES_REQUESTED** [review]

File: js/tbl_change.js [hunk 8]
```diff
@@ -461,6 +461,11 @@ AJAX.registerOnload('tbl_change.js', function () {
      * Continue Insertion form
      */
     $(document).on('change', '#insert_rows', function (event) {
+
+        /**
+         * Displays warning if number of rows are reduced
+         * and chances of data loss.
+         */
```
Comment (mauriciofauth) [inline js/tbl_change.js]: Is this comment really necessary?

Comment (shucon) [reply js/tbl_change.js]: I left it by mistake , thanks.

File: js/tbl_change.js (hunk 6 above)
Comment (shucon) [reply js/tbl_change.js]: @mauriciofauth this jQuery call is used in ajax.js too without variable

**Review by MauricioFauth — APPROVED** [review]

**Comment by mauriciofauth** [comment]
Squashed and merged.
Thanks for your contribution and patience with the feedback.
